internal/indexer/indexer_utils.go

```diff
@@ -0,0 +1,94 @@
+package indexer
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/cometbft/cometbft/state/indexer"
+)
+
+// If the actual event value is a float, we get the condition and parse it as a float
+// to compare against
+func compareFloat(op1 *big.Float, op2 interface{}) (int, bool, error) {
+	switch opVal := op2.(type) {
+	case *big.Int:
+		vF, _, err := big.ParseFloat(opVal.String(), 10, op1.Prec(), big.ToNearestEven)
+		if err != nil {
+			err = fmt.Errorf("failed to convert %s to float", opVal)
+		}
+		cmp := op1.Cmp(vF)
+
+		return cmp, false, err
+
+	case *big.Float:
+		return op1.Cmp(opVal), true, nil
+	default:
+		return -1, false, fmt.Errorf("unable to parse arguments")
+	}
+}
+
+// If the event value we compare against the condition (op2) is an integer
+// we convert the int to float with a precision equal to the number of bits
+// needed to represent the integer to avoid rounding issues with floats
+// where 100 would equal to 100.2 because 100.2 is rounded to 100, while 100.7
+// would be rounded to 101.
+func compareInt(op1 *big.Int, op2 interface{}) (int, bool, error) {
+	switch opVal := op2.(type) {
+	case *big.Int:
+		return op1.Cmp(opVal), false, nil
+	case *big.Float:
+		vF, _, err := big.ParseFloat(op1.String(), 10, uint(op1.BitLen()), big.ToNearestEven)
```

Comment:
It should be `opVal.BitLen()` in both. If the precision is different from that of op1, it will be handled properly in the comparison. Will fix this.

Reply:
I think the idea is to be as precise as needed to compare with `op1` so if `op1` is float then `Prec` is used and if `op1` is Int then `bitten` is used. But as in `compareFloat` you need to `+1`.

Reply:
So, if `op1` is a `float` then we will call `compareFloat`, not `compareInt`.  My point here is that we are converting `op1` to a `float` and we should use as many bits as needed to represent `op1` (whose value is `opVal`). And if the precision of `vF` and `op2` differ, the `Cmp` function will compare them properly (as it compares mantissa's bit by bit and the exponents separetely). So we should not worry about how closely precisions of op1 and op2 match.  
And a `+1` would not necessarily solve it because we might have an int whose precision is bigger than the float in big float so we would need to find the max of `op1.BitLen()` and `op2.Prec()`.

Reply:
Shouldn't there be a way in `math/big` to convert integers to floats without going through a string representation? If my reading of the documentation on [`SetInt`](https://pkg.go.dev/math/big#Float.SetInt) is correct, it's enough to create a `Float` variable and set its precision to 0 (or `op1.BitLen()`, but that will be used under the hood), so that `SetInt` can be used to losslessly convert the value of `op1`?

Reply:
You are absolutely right! Thank you for this, I submitted the updated version. I thought the`setInt` was setting it to the value of an `int64`.